File: parsers/inter/credit_card_csv.py
"""Parser da fatura de cartão de crédito do Banco Inter em CSV — único
formato exportável pro cartão (diferente do extrato da conta corrente).

Formato observado (amostra real, 2026-09/2026-10):

    "Data","Lançamento","Categoria","Tipo","Valor"
    "27/08/2026","SL MARECHAL","COMPRAS","Compra à vista","R$ 32,80"
    "09/09/2026","MLP    *KaBuM-KaBuM","COMPRAS","Parcela 1/10","R$ 133,33"
    "12/08/2026","PAGTO DEBITO AUTOMATICO","OUTROS","Compra à vista","-R$ 50,00"

CSV com vírgula e campos entre aspas (ao contrário do extrato da conta
corrente, que usa ";"). A coluna "Categoria" é a categorização própria do
Inter — ignorada aqui em favor de guess_category (mesma lógica de
palavra-chave usada em todo o app). Linhas com valor negativo são o
pagamento automático da fatura em si (não uma compra) e são descartadas —
o app já tem seu próprio fluxo de "Pagar Fatura", importar essa linha
duplicaria o pagamento.

A coluna "Tipo" indica "Compra à vista" ou "Parcela N/M" — cada parcela de
uma compra parcelada chega como sua própria linha, uma por fatura/mês (o
Inter não manda as parcelas futuras de uma vez, nem um ID que ligue as
parcelas de uma mesma compra entre faturas de meses diferentes). Por isso
não dá pra reconstruir a compra parcelada "de verdade" (como o fluxo manual
"🧾 Compra parcelada" faz, com card_purchase_id agrupando tudo) — o que dá
pra fazer com segurança é só marcar "N/M" na descrição de cada parcela
importada, pra ficar visível qual parcela é e de quantas.
"""
import csv
import io
import re
from datetime import datetime
from typing import List, Optional, Tuple

from parsers.base import NormalizedRow, guess_category
from parsers.inter.common import clean_description, decode_bytes, parse_brl_amount, strip_accents

_INSTALLMENT_RE = re.compile(r"^parcela\s+(\d+)\s*/\s*(\d+)$", re.IGNORECASE)


def _parse_installment(tipo: str) -> Optional[Tuple[int, int]]:
    m = _INSTALLMENT_RE.match(tipo.strip())
    if not m:
        return None
    return int(m.group(1)), int(m.group(2))


class InterCreditCardCsvParser:
    bank_id   = "inter"
    format_id = "credit_card_csv"

    def sniff(self, data: bytes, filename: str) -> bool:
        if not filename.lower().endswith(".csv"):
            return False
        text = decode_bytes(data[:4096])
        upper = strip_accents(text).upper()
        return "LANCAMENTO" in upper and "CATEGORIA" in upper and "VALOR" in upper and '","' in text[:200]

    def parse(self, data: bytes) -> List[NormalizedRow]:
        text = decode_bytes(data)
        reader = csv.reader(io.StringIO(text))
        rows_raw = [r for r in reader if r]
        if not rows_raw:
            return []

        header = [strip_accents(h).strip().upper() for h in rows_raw[0]]
        try:
            i_date = header.index("DATA")
            i_desc = header.index("LANCAMENTO")
            i_val  = header.index("VALOR")
        except ValueError:
            return []
        try:
            i_tipo = header.index("TIPO")
        except ValueError:
            i_tipo = -1

        rows: List[NormalizedRow] = []
        for parts in rows_raw[1:]:
            if len(parts) <= max(i_date, i_desc, i_val):
                continue
            try:
                d = datetime.strptime(parts[i_date].strip(), "%d/%m/%Y").date()
            except ValueError:
                continue

            amount = parse_brl_amount(parts[i_val])
            if amount < 0:
                # Pagamento automático da fatura — não é uma compra, o app
                # já tem seu próprio fluxo de "Pagar Fatura" pra isso.
                continue

            desc = clean_description(parts[i_desc])
            if 0 <= i_tipo < len(parts):
                installment = _parse_installment(parts[i_tipo])
                if installment:
                    n, total = installment
                    desc = f"{desc} (parcela {n}/{total})"

            rows.append(NormalizedRow(
                date=d,
                description=desc,
                amount=amount,
                direction="saida",
                suggested_category=guess_category(desc),
                suggested_payment_method="credito",
                is_credit_card_charge=True,
                raw=",".join(parts),
            ))
        return rows

File: web/src/lib/parsers/inter/creditCardCsv.ts
// Parser da fatura de cartão de crédito do Banco Inter em CSV — único
// formato exportável pro cartão (diferente do extrato da conta corrente).
// Mesma lógica de parsers/inter/credit_card_csv.py (desktop) — ver ali
// para detalhes do formato observado.
//
// A coluna "Tipo" indica "Compra à vista" ou "Parcela N/M" — cada parcela
// de uma compra parcelada chega como sua própria linha, uma por fatura/mês
// (o Inter não manda as parcelas futuras de uma vez, nem um ID que ligue as
// parcelas de uma mesma compra entre faturas de meses diferentes). Por isso
// não dá pra reconstruir a compra parcelada "de verdade" (como o fluxo
// manual "🧾 Compra parcelada" faz, com card_purchase_id agrupando tudo) —
// o que dá pra fazer com segurança é só marcar "N/M" na descrição de cada
// parcela importada, pra ficar visível qual parcela é e de quantas.
import type { BankParser, NormalizedRow } from '../types'
import { guessCategory } from '../base'
import { cleanDescription, decodeBytes, parseBrlAmount, stripAccents } from '../common'

const INSTALLMENT_RE = /^parcela\s+(\d+)\s*\/\s*(\d+)$/i

function parseInstallment(tipo: string): [number, number] | null {
  const m = INSTALLMENT_RE.exec(tipo.trim())
  if (!m) return null
  return [Number(m[1]), Number(m[2])]
}

function parseDateBr(s: string): string | null {
  const m = s.trim().match(/^(\d{2})\/(\d{2})\/(\d{4})$/)
  if (!m) return null
  return `${m[3]}-${m[2]}-${m[1]}`
}

// Parser CSV simples com suporte a campos entre aspas (o formato do Inter
// não tem vírgulas dentro de campos, mas isso lida com aspas corretamente
// mesmo assim, sem precisar de uma lib externa).
function parseCsvLine(line: string): string[] {
  const out: string[] = []
  let cur = ''
  let inQuotes = false
  for (let i = 0; i < line.length; i++) {
    const ch = line[i]
    if (inQuotes) {
      if (ch === '"') {
        if (line[i + 1] === '"') { cur += '"'; i++ } else { inQuotes = false }
      } else {
        cur += ch
      }
    } else if (ch === '"') {
      inQuotes = true
    } else if (ch === ',') {
      out.push(cur)
      cur = ''
    } else {
      cur += ch
    }
  }
  out.push(cur)
  return out
}

export class InterCreditCardCsvParser implements BankParser {
  bankId = 'inter'
  formatId = 'credit_card_csv'

  sniff(bytes: ArrayBuffer, filename: string): boolean {
    if (!filename.toLowerCase().endsWith('.csv')) return false
    const text = decodeBytes(bytes.slice(0, 4096))
    const upper = stripAccents(text).toUpperCase()
    return upper.includes('LANCAMENTO') && upper.includes('CATEGORIA') && upper.includes('VALOR')
      && text.slice(0, 200).includes('","')
  }

  async parse(bytes: ArrayBuffer): Promise<NormalizedRow[]> {
    const text = decodeBytes(bytes)
    const lines = text.split(/\r?\n/).filter((ln) => ln.trim())
    if (lines.length === 0) return []

    const header = parseCsvLine(lines[0]).map((h) => stripAccents(h).trim().toUpperCase())
    const iDate = header.indexOf('DATA')
    const iDesc = header.indexOf('LANCAMENTO')
    const iVal = header.indexOf('VALOR')
    const iTipo = header.indexOf('TIPO')
    if (iDate === -1 || iDesc === -1 || iVal === -1) return []

    const rows: NormalizedRow[] = []
    for (const ln of lines.slice(1)) {
      const parts = parseCsvLine(ln)
      if (parts.length <= Math.max(iDate, iDesc, iVal)) continue

      const isoDate = parseDateBr(parts[iDate])
      if (!isoDate) continue

      const amount = parseBrlAmount(parts[iVal])
      if (amount < 0) {
        // Pagamento automático da fatura — não é uma compra, o app já tem
        // seu próprio fluxo de "Pagar Fatura" pra isso.
        continue
      }

      let desc = cleanDescription(parts[iDesc])
      if (iTipo !== -1 && iTipo < parts.length) {
        const installment = parseInstallment(parts[iTipo])
        if (installment) {
          const [n, total] = installment
          desc = `${desc} (parcela ${n}/${total})`
        }
      }
      rows.push({
        date: isoDate,
        description: desc,
        amount,
        direction: 'saida',
        suggestedCategory: guessCategory(desc),
        suggestedPaymentMethod: 'credito',
        isInvestmentLike: false,
        isCreditCardCharge: true,
        raw: ln,
      })
    }
    return rows
  }
}
